File: research/gaia/gremlin/gremlin_core/tests/filter_test.rs
//
//! Copyright 2020 Alibaba Group Holding Limited.
//!
//! Licensed under the Apache License, Version 2.0 (the "License");
//! you may not use this file except in compliance with the License.
//! You may obtain a copy of the License at
//!
//! http://www.apache.org/licenses/LICENSE-2.0
//!
//! Unless required by applicable law or agreed to in writing, software
//! distributed under the License is distributed on an "AS IS" BASIS,
//! WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//! See the License for the specific language governing permissions and
//! limitations under the License.

mod common;

#[cfg(test)]
mod test {
    use crate::common::test::*;
    use gremlin_core::process::traversal::traverser::Requirement;

    // g.V().hasLabel("PERSON")
    #[test]
    fn has_step_test_01() {
        initialize();
        let mut expected = to_global_ids(vec![1, 2, 4, 6]);
        expected.sort();
        let test_job_factory = TestJobFactory::with_expect_ids(expected);
        let pb_request = read_pb_request(gen_path("has_step_test_01")).expect("read pb failed");
        run_test(test_job_factory, pb_request);
    }

    // g.V().hasId(1)
    #[test]
    fn has_step_test_02() {
        initialize();
        let mut expected = to_global_ids(vec![1]);
        expected.sort();
        let test_job_factory = TestJobFactory::with_expect_ids(expected);
        let pb_request = read_pb_request(gen_path("has_step_test_02")).expect("read pb failed");
        run_test(test_job_factory, pb_request);
    }

    // g.V().has("name", "marko")
    #[test]
    fn has_step_test_03() {
        initialize();
        let mut expected = to_global_ids(vec![1]);
        expected.sort();
        let test_job_factory = TestJobFactory::with_expect_ids(expected);
        let pb_request = read_pb_request(gen_path("has_step_test_03")).expect("read pb failed");
        run_test(test_job_factory, pb_request);
    }

    // g.V().has("id", neq(1))
    #[test]
    fn has_step_test_04() {
        initialize();
        let mut expected = to_global_ids(vec![2, 3, 4, 5, 6]);
        expected.sort();
        let test_job_factory = TestJobFactory::with_expect_ids(expected);
        let pb_request = read_pb_request(gen_path("has_step_test_04")).expect("read pb failed");
        run_test(test_job_factory, pb_request);
    }

    // g.V().has("age", lte(28).or(gte(32)))
    #[test]
    fn has_step_test_05() {
        initialize();
        let mut expected = to_global_ids(vec![2, 4, 6]);
        expected.sort();
        let test_job_factory = TestJobFactory::with_expect_ids(expected);
        let pb_request = read_pb_request(gen_path("has_step_test_05")).expect("read pb failed");
        run_test(test_job_factory, pb_request);
    }

    // g.V().has("age", inside(28,32))
    #[test]
    fn has_step_test_06() {
        initialize();
        let mut expected = to_global_ids(vec![1]);
        expected.sort();
        let test_job_factory = TestJobFactory::with_expect_ids(expected);
        let pb_request = read_pb_request(gen_path("has_step_test_06")).expect("read pb failed");
        run_test(test_job_factory, pb_request);
    }

    // g.V().has("age", P.without(27))
    #[test]
    fn has_step_test_07() {
        initialize();
        let mut expected = to_global_ids(vec![1, 4, 6]);
        expected.sort();
        let test_job_factory = TestJobFactory::with_expect_ids(expected);
        let pb_request = read_pb_request(gen_path("has_step_test_07")).expect("read pb failed");
        run_test(test_job_factory, pb_request);
    }

    // g.V().has("age", P.within(27))
    #[test]
    fn has_step_test_08() {
        initialize();
        let mut expected = to_global_ids(vec![2]);
        expected.sort();
        let test_job_factory = TestJobFactory::with_expect_ids(expected);
        let pb_request = read_pb_request(gen_path("has_step_test_08")).expect("read pb failed");
        run_test(test_job_factory, pb_request);
    }

    // g.V().as("a").both("KNOWS").where(both("KNOWS").as("a"))
    #[test]
    fn where_step_test_01() {
        initialize();
        let mut expected = to_global_ids(vec![1, 1, 2, 4]);
        expected.sort();
        let mut test_job_factory = TestJobFactory::with_expect_ids(expected);
        test_job_factory.set_requirement(Requirement::LABELED_PATH);
        let pb_request = read_pb_request(gen_path("where_step_test_01")).expect("read pb failed");
        run_test(test_job_factory, pb_request);
    }

    // g.V().as("a").out("created").where(__.as("a").values("name").is("josh")).in("created")
    #[test]
    fn where_step_test_02() {
        initialize();
        let mut expected = to_global_ids(vec![1, 4, 4, 6]);
        expected.sort();
        let mut test_job_factory = TestJobFactory::with_expect_ids(expected);
        test_job_factory.set_requirement(Requirement::LABELED_PATH);
        let pb_request = read_pb_request(gen_path("where_step_test_02")).expect("read pb failed");
        run_test(test_job_factory, pb_request);
    }

    // g.V().as("a").out("created").in("created").as("b").where("a", P.gt("b")).by("age")
    #[test]
    fn where_step_test_03() {
        initialize();
        let mut expected = to_global_ids(vec![1, 1, 4]);
        expected.sort();
        let mut test_job_factory = TestJobFactory::with_expect_ids(expected);
        test_job_factory.set_requirement(Requirement::LABELED_PATH);
        let pb_request = read_pb_request(gen_path("where_step_test_03")).expect("read pb failed");
        run_test(test_job_factory, pb_request);
    }

    // g.V().count().unfold().is(6)
    #[test]
    fn is_step_test_01() {
        initialize();
        let expected = vec![6.into()];
        let test_job_factory = TestJobFactory::with_expect_values(expected);
        let pb_request = read_pb_request(gen_path("is_step_test_01")).expect("read pb failed");
        run_test(test_job_factory, pb_request);
    }

    // g.V().count().unfold().is(neq(6))
    #[test]
    fn is_step_test_02() {
        initialize();
        let expected = vec![];
        let test_job_factory = TestJobFactory::with_expect_values(expected);
        let pb_request = read_pb_request(gen_path("is_step_test_02")).expect("read pb failed");
        run_test(test_job_factory, pb_request);
    }

    // g.V().count().unfold().is(gt(5))
    #[test]
    fn is_step_test_03() {
        initialize();
        let expected = vec![6.into()];
        let test_job_factory = TestJobFactory::with_expect_values(expected);
        let pb_request = read_pb_request(gen_path("is_step_test_03")).expect("read pb failed");
        run_test(test_job_factory, pb_request);
    }

    // g.V().count().unfold().is(lt(7))
    #[test]
    fn is_step_test_04() {
        initialize();
        let expected = vec![6.into()];
        let test_job_factory = TestJobFactory::with_expect_values(expected);
        let pb_request = read_pb_request(gen_path("is_step_test_04")).expect("read pb failed");
        run_test(test_job_factory, pb_request);
    }
}

File: interactive_engine/compiler/src/test/java/com/alibaba/graphscope/common/ir/FilterTest.java
/*
 * Copyright 2020 Alibaba Group Holding Limited.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.alibaba.graphscope.common.ir;

import com.alibaba.graphscope.common.exception.FrontendException;
import com.alibaba.graphscope.common.ir.tools.GraphBuilder;
import com.alibaba.graphscope.common.ir.tools.GraphRexBuilder;
import com.alibaba.graphscope.common.ir.tools.GraphStdOperatorTable;
import com.alibaba.graphscope.common.ir.tools.config.GraphOpt;
import com.alibaba.graphscope.common.ir.tools.config.LabelConfig;
import com.alibaba.graphscope.common.ir.tools.config.SourceConfig;
import com.alibaba.graphscope.common.ir.type.GraphProperty;
import com.google.common.collect.ImmutableList;

import org.apache.calcite.rel.RelNode;
import org.apache.calcite.rel.logical.LogicalValues;
import org.apache.calcite.rex.RexBuilder;
import org.apache.calcite.rex.RexCall;
import org.apache.calcite.rex.RexNode;
import org.apache.calcite.sql.type.SqlTypeName;
import org.junit.Assert;
import org.junit.Test;

public class FilterTest {
    // source([person]).filter("XXX") are fused
    @Test
    public void equal_1_test() {
        GraphBuilder builder = Utils.mockGraphBuilder();
        SourceConfig sourceConfig =
                new SourceConfig(GraphOpt.Source.VERTEX, new LabelConfig(false).addLabel("person"));
        RexNode equal =
                builder.source(sourceConfig)
                        .call(
                                GraphStdOperatorTable.EQUALS,
                                builder.variable(null, "age"),
                                builder.literal(10));
        RelNode filter = builder.filter(equal).build();
        Assert.assertEquals(
                "GraphLogicalSource(tableConfig=[{isAll=false, tables=[person]}], alias=[_],"
                        + " fusedFilter=[[=(_.age, 10)]], opt=[VERTEX])",
                filter.explain().trim());
    }

    // source([person]).as('x').filter("XXX") are fused
    @Test
    public void equal_2_test() {
        GraphBuilder builder = Utils.mockGraphBuilder();
        SourceConfig sourceConfig =
                new SourceConfig(
                        GraphOpt.Source.VERTEX, new LabelConfig(false).addLabel("person"), "x");
        RexNode equal =
                builder.source(sourceConfig)
                        .call(
                                GraphStdOperatorTable.EQUALS,
                                builder.variable(null, "age"),
                                builder.literal(10));
        RelNode filter = builder.filter(equal).build();
        Assert.assertEquals(
                "GraphLogicalSource(tableConfig=[{isAll=false, tables=[person]}], alias=[x],"
                        + " fusedFilter=[[=(_.age, 10)]], opt=[VERTEX])",
                filter.explain().trim());
    }

    // source([person]).as('x').filter('x.age == 10') are fused
    @Test
    public void equal_3_test() {
        GraphBuilder builder = Utils.mockGraphBuilder();
        SourceConfig sourceConfig =
                new SourceConfig(
                        GraphOpt.Source.VERTEX, new LabelConfig(false).addLabel("person"), "x");
        RexNode equal =
                builder.source(sourceConfig)
                        .call(
                                GraphStdOperatorTable.EQUALS,
                                builder.variable("x", "age"),
                                builder.literal(10));
        RelNode filter = builder.filter(equal).build();
        Assert.assertEquals(
                "GraphLogicalSource(tableConfig=[{isAll=false, tables=[person]}], alias=[x],"
                        + " fusedFilter=[[=(_.age, 10)]], opt=[VERTEX])",
                filter.explain().trim());
    }

    // g.V().hasLabel("person").where(expr("@.age > 10"))
    @Test
    public void greater_1_test() {
        GraphBuilder builder = Utils.mockGraphBuilder();
        SourceConfig sourceConfig =
                new SourceConfig(GraphOpt.Source.VERTEX, new LabelConfig(false).addLabel("person"));
        RexNode greater =
                builder.source(sourceConfig)
                        .call(
                                GraphStdOperatorTable.GREATER_THAN,
                                builder.variable(null, "age"),
                                builder.literal(10));
        RelNode filter = builder.filter(greater).build();
        Assert.assertEquals(
                "GraphLogicalSource(tableConfig=[{isAll=false, tables=[person]}], alias=[_],"
                        + " fusedFilter=[[>(_.age, 10)]], opt=[VERTEX])",
                filter.explain().trim());
    }

    // 20 > 10 -> always returns true, ignore the condition
    @Test
    public void greater_2_test() {
        GraphBuilder builder = Utils.mockGraphBuilder();
        SourceConfig sourceConfig =
                new SourceConfig(GraphOpt.Source.VERTEX, new LabelConfig(false).addLabel("person"));
        RexNode greater =
                builder.source(sourceConfig)
                        .call(
                                GraphStdOperatorTable.GREATER_THAN,
                                builder.literal(20),
                                builder.literal(10));
        RelNode filter = builder.filter(greater).build();
        Assert.assertEquals(
                "GraphLogicalSource(tableConfig=[{isAll=false, tables=[person]}], alias=[_],"
                        + " opt=[VERTEX])",
                filter.explain().trim());
    }

    /**
     * 10 > 20 -> always returns false, create {@link LogicalValues} which carries all data types of the node before the filter
     */
    @Test
    public void greater_3_test() {
        GraphBuilder builder = Utils.mockGraphBuilder();
        SourceConfig sourceConfig =
                new SourceConfig(GraphOpt.Source.VERTEX, new LabelConfig(false).addLabel("person"));
        RexNode greater =
                builder.source(sourceConfig)
                        .call(
                                GraphStdOperatorTable.GREATER_THAN,
                                builder.literal(10),
                                builder.literal(20));
        // the node before the filter
        RelNode previous = builder.peek();
        RelNode filter = builder.filter(greater).build();
        Assert.assertEquals(filter.getClass(), LogicalValues.class);
        Assert.assertEquals(filter.getRowType(), previous.getRowType());
    }

    // test fuzzy conditions: g.V().has("age", gt(10))
    @Test
    public void greater_4_test() {
        GraphBuilder builder = Utils.mockGraphBuilder();
        SourceConfig sourceConfig = new SourceConfig(GraphOpt.Source.VERTEX, new LabelConfig(true));
        RexNode greater =
                builder.source(sourceConfig)
                        .call(
                                GraphStdOperatorTable.GREATER_THAN,
                                builder.variable(null, "age"),
                                builder.literal(10));
        RelNode filter = builder.filter(greater).build();
        Assert.assertEquals(
                "GraphLogicalSource(tableConfig=[{isAll=true, tables=[software, person]}],"
                        + " alias=[_], fusedFilter=[[>(_.age, 10)]], opt=[VERTEX])",
                filter.explain().trim());
    }

    // Match (:person {age > $age})
    @Test
    public void greater_5_test() {
        GraphBuilder builder = Utils.mockGraphBuilder();
        SourceConfig sourceConfig =
                new SourceConfig(GraphOpt.Source.VERTEX, new LabelConfig(false).addLabel("person"));
        RexCall greater =
                (RexCall)
                        builder.source(sourceConfig)
                                .call(
                                        GraphStdOperatorTable.GREATER_THAN,
                                        builder.variable(null, "age"),
                                        ((GraphRexBuilder) builder.getRexBuilder())
                                                .makeGraphDynamicParam("age", 0));
        Assert.assertEquals(">(_.age, ?0)", greater.toString());
        Assert.assertEquals(
                SqlTypeName.INTEGER, greater.getOperands().get(1).getType().getSqlTypeName());
    }

    // g.V().hasLabel("person").where(expr("@.age > 20 and @.name == marko"))
    @Test
    public void and_1_test() {
        GraphBuilder builder = Utils.mockGraphBuilder();
        SourceConfig sourceConfig =
                new SourceConfig(GraphOpt.Source.VERTEX, new LabelConfig(false).addLabel("person"));
        RexNode condition1 =
                builder.source(sourceConfig)
                        .call(
                                GraphStdOperatorTable.GREATER_THAN,
                                builder.variable(null, "age"),
                                builder.literal(20));
        RexNode condition2 =
                builder.call(
                        GraphStdOperatorTable.EQUALS,
                        builder.variable(null, "name"),
                        builder.literal("marko"));
        RelNode filter = builder.filter(condition1, condition2).build();
        Assert.assertEquals(
                "GraphLogicalSource(tableConfig=[{isAll=false, tables=[person]}], alias=[_],"
                        + " fusedFilter=[[AND(>(_.age, 20), =(_.name, _UTF-8'marko'))]],"
                        + " opt=[VERTEX])",
                filter.explain().trim());
    }

    // g.V().hasLabel("person").where(expr("@.age > 20 and @.age < 30"))
    @Test
    public void and_2_test() {
        GraphBuilder builder = Utils.mockGraphBuilder();
        SourceConfig sourceConfig =
                new SourceConfig(GraphOpt.Source.VERTEX, new LabelConfig(false).addLabel("person"));
        RexNode condition1 =
                builder.source(sourceConfig)
                        .call(
                                GraphStdOperatorTable.GREATER_THAN,
                                builder.variable(null, "age"),
                                builder.literal(20));
        RexNode condition2 =
                builder.call(
                        GraphStdOperatorTable.LESS_THAN,
                        builder.variable(null, "age"),
                        builder.literal(30));
        RelNode filter =
                builder.filter(builder.call(GraphStdOperatorTable.AND, condition1, condition2))
                        .build();
        Assert.assertEquals(
                "GraphLogicalSource(tableConfig=[{isAll=false, tables=[person]}], alias=[_],"
                        + " fusedFilter=[[AND(>(_.age, 20), <(_.age, 30))]], opt=[VERTEX])",
                filter.explain().trim());
    }

    // g.V().hasLabel('person')
    @Test
    public void label_equals_1_test() {
        GraphBuilder builder = Utils.mockGraphBuilder();
        RelNode node =
                builder.source(new SourceConfig(GraphOpt.Source.VERTEX))
                        .filter(
                                builder.call(
                                        GraphStdOperatorTable.EQUALS,
                                        builder.variable(null, GraphProperty.LABEL_KEY),
                                        builder.literal("person")))
                        .build();
        Assert.assertEquals(
                "GraphLogicalSource(tableConfig=[{isAll=false, tables=[person]}], alias=[_],"
                        + " opt=[VERTEX])",
                node.explain().trim());
    }

    // g.V().hasLabel('XX'), 'XX' is not a valid label
    @Test
    public void label_equals_2_test() {
        try {
            GraphBuilder builder = Utils.mockGraphBuilder();
            RelNode node =
                    builder.source(new SourceConfig(GraphOpt.Source.VERTEX))
                            .filter(
                                    builder.call(
                                            GraphStdOperatorTable.EQUALS,
                                            builder.variable(null, GraphProperty.LABEL_KEY),
                                            builder.literal("XX")))
                            .build();
        } catch (IllegalArgumentException e) {
            Assert.assertEquals(
                    "cannot find common labels between values= [XX] and label="
                            + " [[VertexLabel(software), VertexLabel(person)]]",
                    e.getMessage());
            return;
        }
        Assert.fail();
    }

    // g.V().hasLabel('person').hasLabel('software')
    @Test
    public void label_equals_3_test() {
        try {
            GraphBuilder builder = Utils.mockGraphBuilder();
            RelNode node =
                    builder.source(
                                    new SourceConfig(
                                            GraphOpt.Source.VERTEX,
                                            new LabelConfig(false).addLabel("person")))
                            .filter(
                                    builder.call(
                                            GraphStdOperatorTable.EQUALS,
                                            builder.variable(null, GraphProperty.LABEL_KEY),
                                            builder.literal("software")))
                            .build();
        } catch (IllegalArgumentException e) {
            Assert.assertEquals(
                    "cannot find common labels between values= [software] and label="
                            + " [[VertexLabel(person)]]",
                    e.getMessage());
            return;
        }
        Assert.fail();
    }

    // g.V().has('age', 10).hasLabel('person')
    @Test
    public void label_equals_4_test() {
        GraphBuilder builder = Utils.mockGraphBuilder();
        RelNode node =
                builder.source(new SourceConfig(GraphOpt.Source.VERTEX))
                        .filter(
                                builder.call(
                                        GraphStdOperatorTable.EQUALS,
                                        builder.variable(null, "age"),
                                        builder.literal(10)))
                        .filter(
                                builder.call(
                                        GraphStdOperatorTable.EQUALS,
                                        builder.variable(null, GraphProperty.LABEL_KEY),
                                        builder.literal("person")))
                        .build();
        Assert.assertEquals(
                "GraphLogicalSource(tableConfig=[{isAll=false, tables=[person]}], alias=[_],"
                        + " fusedFilter=[[=(_.age, 10)]], opt=[VERTEX])",
                node.explain().trim());
    }

    // g.V().has('age', 10).hasLabel('software'), property 'age' not exist in label='software'
    @Test
    public void label_equals_5_test() {
        try {
            GraphBuilder builder = Utils.mockGraphBuilder();
            RelNode node =
                    builder.source(new SourceConfig(GraphOpt.Source.VERTEX))
                            .filter(
                                    builder.call(
                                            GraphStdOperatorTable.EQUALS,
                                            builder.variable(null, "age"),
                                            builder.literal(10)))
                            .filter(
                                    builder.call(
                                            GraphStdOperatorTable.EQUALS,
                                            builder.variable(null, GraphProperty.LABEL_KEY),
                                            builder.literal("software")))
                            .build();
        } catch (FrontendException e) {
            Assert.assertTrue(
                    e.getMessage()
                            .contains(
                                    "{property=age} not found; expected properties are: [id, name,"
                                            + " lang, creationDate]"));
            return;
        }
        Assert.fail();
    }

    // ~label within ['person']
    @Test
    public void label_within_1_test() {
        GraphBuilder builder = Utils.mockGraphBuilder();
        RexBuilder rexBuilder = builder.getRexBuilder();
        RelNode node =
                builder.source(new SourceConfig(GraphOpt.Source.VERTEX))
                        .filter(
                                rexBuilder.makeIn(
                                        builder.variable(null, GraphProperty.LABEL_KEY),
                                        ImmutableList.of(builder.literal("person"))))
                        .build();
        Assert.assertEquals(
                "GraphLogicalSource(tableConfig=[{isAll=false, tables=[person]}], alias=[_],"
                        + " opt=[VERTEX])",
                node.explain().trim());
    }

    // g.V().hasId(1, 2, 3)
    @Test
    public void unique_key_within_test() {
        GraphBuilder builder = Utils.mockGraphBuilder();
        RexBuilder rexBuilder = builder.getRexBuilder();
        RelNode node =
                builder.source(new SourceConfig(GraphOpt.Source.VERTEX))
                        .filter(
                                rexBuilder.makeIn(
                                        builder.variable(null, GraphProperty.ID_KEY),
                                        ImmutableList.of(
                                                builder.literal(1),
                                                builder.literal(2),
                                                builder.literal(3))))
                        .build();
        Assert.assertEquals(
                "GraphLogicalSource(tableConfig=[{isAll=true, tables=[software, person]}],"
                        + " alias=[_], opt=[VERTEX], uniqueKeyFilters=[SEARCH(_.~id,"
                        + " Sarg[1, 2, 3])])",
                node.explain().trim());
    }
}
